odl/space/space_utils.py

```diff
@@ -203,6 +206,233 @@ def rn(size, dtype=None, impl='numpy', **kwargs):
     return rn
 
 
+class auto_weighting(OptionalArgDecorator):
+
+    """Make an unweighted adjoint automatically account for weightings.
+
+    Depending on the weightings, the correction is achieved by composing
+    the unweighted operator with either `ScalingOperator` or
+    `ConstantOperator`. The following rules are applied for the domain
+    weighting ``w``, the range weighting ``v`` and the provided unweighted
+    adjoint ``B^*``:
+
+    - If both ``w`` and ``v`` are arrays, return ::
+
+        (1 / w) * (B^*) * v
+
+    - If ``w`` is an array and ``v`` a constant, return ::
+
+        (v / w) * (B^*)
+
+    - If ``w`` is a constant and ``v`` an array, return ::
+
+        (B^*) * (w / v)
+
+    - If both ``w`` and ``v`` are constants, return ::
+
+        (B^*) * (v / w)
+
+      if ``B.range.size < B.domain.size``, otherwise ::
+
+        (v / w) * (B^*)
+
+    - Ignore constants 1.0.
+
+    To avoid the inconvenience of dealing with `OperatorComp` objects,
+    the given operator is monkey-patched instead of composed.
+
+    Parameters
+    ----------
+    unweighted_adjoint : `Operator`
+        Unweighted variant of the adjoint. It will be patched with a
+        new ``_call()`` method.
+        The weightings of ``domain`` and ``range`` of the operator
+        must be `ArrayWeighting` or `ConstWeighting`.
+    optimize : bool, optional
+        If ``True``, merge and move around constant weightings for
+        highest expected efficiency.
+
+    Notes
+    -----
+    Consider a linear operator :math:`A: X_w \\to Y_v` between spaces with
+    weights :math:`w` and :math:`v`, respectively, along with the same
+    operator :math:`B: X \\to Y` defined between the unweighted variants of
+    the spaces. (This means that :math:`B f = A f` for all
+    :math:`f \\in X \cong X_w`).
+
+    Then, the adjoint of :math:`A` is related to the adjoint of :math:`B`
+    as follows:
+
+    .. math::
+        \\langle Af, g \\rangle_{Y_v} =
+        \\langle Bf, v \cdot g \\rangle_Y =
+        \\langle f, B^*(v \cdot g) \\rangle_X =
+        \\langle f, w^{-1}\, B^*(v \cdot g) \\rangle_{X_w}.
+
+    Thus, from the existing unweighted adjoint :math:`B^*` one can compute
+    the weighted one as :math:`A^* = w^{-1}\, B^*(v\, \cdot)`.
+    Depending on the types of weighting, this expression can be simplified
+    further, e.g., a constant weight can be absorbed into the other weight.
+    """
```

Comment:
This could use a minimal example

Reply:
If you refer to the merging rules, they are all explained in detail. As for the general idea, I agree.